ENH: zero_division parameter for classification…

marctorsoc opened the pull request:
See issue #14876 

#### What does this implement/fix? Explain your changes.

zero_division parameter for precision, recall, and friends

#### Any other comments?

3 possible values: 
* "warn": same behavior as before
* 0, 1: remove the warnings and set value for the metrics to this when the metrics is ill-defined (see issue for more details and examples)


Just to clarify:

![image](https://user-images.githubusercontent.com/22045779/65581402-4bf39b00-df73-11e9-9bc9-ef6bae636c7b.png)

prec will be ZD if all predictions are negative
rec will be ZD if all labels are negative
f will be ZD if everything is negative

Note that if ZD = "warn" this means 0 + warning

marctorsoc committed:
first commit for issue 14876: zero_division parameter

jnothman commented on doc/whats_new/v0.21.rst:
```diff
@@ -69,6 +69,13 @@ Support for Python 3.4 and below has been officially dropped.
 ......................
 
+- |Feature| Added a new parameter ``zero_division`` to multiple classification
+metrics: :func:`precision_score`, :func:`recall_score`, :func:`f1_score`,
```
please indent

jnothman commented on sklearn/metrics/classification.py:
```diff
@@ -892,6 +903,12 @@ def f1_score(y_true, y_pred, labels=None, pos_label=1, average='binary',
     sample_weight : array-like of shape = [n_samples], optional
         Sample weights.
 
+    zero_division : string or int, default="warn"
```
```suggestion
    zero_division : "warn", 0 or 1
```

Hmm... thinking about it, f1_score will be 0 in all cases. So this is only about suppressing the warning, if we're interested in such API-level control.

marctorrellas replied:
Please give me some time (I'll do it during the weekend) to review and do another commit. I just merged a very conflicting PR and this needs changes. Regarding the values of f1, we agreed that when y_true and y_pred have none of them positives, we would give the user the option to return 1. Why now this change?

jnothman reviewed (commented):
thanks for opening the PR. Sorry I'm not able to review immediately.

marctorsoc committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into prec_rec_fscore_zero_division

# Conflicts:
#	sklearn/metrics/classification.py
#	sklearn/metrics/tests/test_classification.py

lost some stuff after merging, need a review

jnothman commented:
Oh yes, forgot that.

marctorsoc committed:
- merge with master
- Changed whats_new to 0.22
- F-score only warns if both prec and rec are ill-defined
- new private method to simplify _prf_divide

marctorrellas commented:
> thanks for opening the PR. Sorry I'm not able to review immediately.

Hi @jnothman , it's just my second PR to sklearn so I'm still learning :)

I'm having a problem with git, it says I have 9 files changed, but actually changed only 3. It's like it's comparing with a master from some days ago. For example this commit:

https://github.com/scikit-learn/scikit-learn/commit/2119490c10675839e546abfedba75d30bfbdc8ad

is in master but in this PR its changes appear in the diff. Can you guide me to fix this?

marctorsoc committed:
fixed "[...0 0...]" --> "[...0, 0...]" in docstring

marctorsoc committed:
corrected docstring examples

marctorsoc committed:
corrected docstring examples (again)

jnothman commented on sklearn/metrics/classification.py:
```diff
@@ -892,6 +903,12 @@ def f1_score(y_true, y_pred, labels=None, pos_label=1, average='binary',
 
+    zero_division : string or int, default="warn"
+        Sets the behavior when there is a zero division. If set to
+        ("warn"|0)/1, returns 0/1 when both precision and recall are zero
```
I don't think this notation is easy enough to read. How about 'Sets the value to return when blah blah. If "warn" (default), this acts like `0` but also raises a warning.'

marctorrellas replied:
wrote something similar, please check new version

jnothman commented on sklearn/metrics/classification.py:
```diff
@@ -1062,7 +1092,12 @@ def _prf_divide(numerator, denominator, metric, modifier, average, warn_for):
 
     # remove infs
-    result[mask] = 0.0
+    result[mask] = float(zero_division == 1)
```
this is obfuscated. I'd rather `0.0 if zero_division in ('warn', 0) else 1`

marctorrellas replied:
done in new version

jnothman commented on sklearn/metrics/classification.py:
```diff
@@ -42,6 +42,17 @@
+def _check_zero_division(zero_division):
+    if isinstance(zero_division, str):
+        if zero_division != "warn":
+            raise ValueError('zero_division must be one of ["warn", 0, 1]')
```
Might as well add `f'Got {zero_division!r}'`

marctorrellas replied:
I tried this but f-strings are not supported by all python versions we're testing with. So added with `.format`

jnothman commented on sklearn/metrics/classification.py:
```diff
@@ -42,6 +42,17 @@
+        if zero_division not in [0, 1]:
+            raise ValueError('zero_division must be one of ["warn", 0, 1]')
+    else:
+        raise TypeError('zero_division must be one of ["warn", 0, 1]')
```
we aren't usually pedantic like this in distinguishing TypeErrors for parameter types frrom `ValueErrors`

marctorrellas replied:
simplified in the new version

jnothman commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -1348,26 +1352,34 @@ def test_precision_recall_f1_score_with_an_empty_prediction():
+    fbeta = my_assert(*tmp, y_true, y_pred, beta=beta,
+                      average=average, zero_division=zero_division)
+
+    zero_division = float(zero_division == 1)
```
obfuscated

marctorrellas replied:
simplified in the new version with two separated tests

jnothman commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -1381,117 +1393,167 @@ def test_precision_recall_f1_no_labels_average_none():
-    assert_array_almost_equal(r, [0, 0, 0], 2)
-    assert_array_almost_equal(f, [0, 0, 0], 2)
+    func = precision_recall_fscore_support
+    my_assert = (assert_warns if zero_division == "warn"
```
if you must do something like this, use functools.partial to capture the arguments too.

But I think tests must be very readable code, as the reader needs to be absolutely certain of their correctness to be confident that they in turn imply the corretness of the code.

marctorrellas replied:
simplified in the new version with two separated tests

jnothman commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -1381,117 +1393,167 @@ def test_precision_recall_f1_no_labels_average_none():
+    fbeta = my_assert(*tmp, y_true, y_pred, beta=beta,
+                      average=None, zero_division=zero_division)
+
+    zero_division = float(zero_division == 1)
```
This is obfuscated. I'd rather a clear, separate test checking the behaviour of zero_divison, than a tiny, unexplicit piece in a larger test.

marctorrellas replied:
simplified in the new version with two separated tests

marctorsoc committed:
- tests for warn and [0,1] separated to make them more clear;
- better docstrings
- more explicit use of zero_division value

jnothman commented on doc/developers/advanced_installation.rst:
```diff
@@ -242,12 +242,20 @@ The above commands assume that you have the Python installation folder in your
-<https://visualstudio.microsoft.com/de/downloads/>`_.
+<https://visualstudio.microsoft.com/downloads/>`_.
+
+.. warning::
```
You've done something strange in trying to merge in changes from master. Please try to merge in the latest master again

marctorsoc committed:
Merge branch 'sklearn_master' into prec_rec_fscore_zero_division

# Conflicts:
#	doc/whats_new/v0.22.rst

marctorsoc committed:
Merge branch 'sklearn_master' into prec_rec_fscore_zero_division

marctorrellas commented:
@jnothman any more comments?

marctorsoc committed:
- removed fstring to make compatible with python<3.6

jnothman commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -1549,136 +1607,233 @@ def test_precision_recall_f1_no_labels_average_none():
 
 
-def test_prf_warnings():
+@pytest.mark.parametrize('zero_division', ["warn"])
```
not sure how this helps

jnothman reviewed (commented):
Thanks for the ping.

I don't think we currently test the return value (i.e. zero_division=1) except in the case that all the labels (true and pred) are negative... we don't seem to test zero_division=1 in the zero-sample_weight case either (though it is a pretty weird case).

marctorrellas commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -1514,30 +1516,86 @@ def test_precision_recall_f1_score_with_an_empty_prediction():
 
 
-def test_precision_recall_f1_no_labels_average_none():
+@pytest.mark.parametrize('zero_division', [0, 1])
```
Given your previous comment:

```This is obfuscated. I'd rather a clear, separate test checking the behaviour of zero_divison, than a tiny, unexplicit piece in a larger test.```

I decided to separate this into two tests. I think it is a lot more readable. Otherwise, there are if's or the use of functools.partial. I can go back to previous version, but honestly, if we want readability I think this is better (maybe with better names)

jnothman commented:
Ahh yes, the pytest.warns context manager might help with making the no
warning case fit a similar structure to the yes warning case, using
pytest.warns(None). But okay.

marctorrellas commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -1875,7 +2030,7 @@ def test_hinge_loss_multiclass_with_missing_labels():
     dummy_hinge_loss = np.mean(dummy_losses)
     assert (hinge_loss(y_true, pred_decision, labels=labels) ==
-                 dummy_hinge_loss)
+            dummy_hinge_loss)
```
if I don't change this I have flake8 warning:

```sklearn/metrics/tests/test_classification.py:1988:18: E127 continuation line over-indented for visual indent```

jnothman replied:
Yes, I know that this is bad PEP8... we've considered black, but not clearly decided in its favour

marctorrellas commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -534,10 +534,12 @@ def test_cohen_kappa():
-                        weights="linear"), 0.9412, decimal=4)
-    assert_almost_equal(cohen_kappa_score(y1, y2,
-                        weights="quadratic"), 0.9541, decimal=4)
+    assert_almost_equal(
```
sorry for that. This formatting things are so annoying, have you considered `black`? it's really handy

marctorrellas commented:
update table in the description, it was wrong :(

prec will be ZD if all predictions are negative
rec will be ZD if all labels are negative
f will be ZD if everything is negative

marctorsoc committed:
- reverted changes to avoid flake8 warnings
- added tests for YTN or YPN to check prec/rec with zero_division value
- cleaner tests

marctorrellas commented:
> Thanks for the ping.
> 
> I don't think we currently test the return value (i.e. zero_division=1) except in the case that all the labels (true and pred) are negative... we don't seem to test zero_division=1 in the zero-sample_weight case either (though it is a pretty weird case).

Added zero_division to a test where prec and rec both have their peculiar cases. I don't understand the last comment, do you mean passing the `labels` param with a label that is not present?

jnothman commented:
The last comment was about a special case in the code where
sample_weight=zeros

jnothman reviewed (commented):
Thank you. This is looking good!

Let's see what others think about this, including the parameter name which I think is still up for debate.

jnothman reviewed (approved):
Thank you. This is looking good!

Let's see what others think about this, including the parameter name which I think is still up for debate.

marctorrellas commented:
> Thank you. This is looking good!
> 
> Let's see what others think about this, including the parameter name which I think is still up for debate.

Thanks!

thomasjpfan reviewed (commented):
Would `on_zero_division` be a better name?

marctorrellas commented on sklearn/metrics/classification.py:
```diff
@@ -41,6 +41,17 @@
 
 
+def _check_zero_division(zero_division):
+    if isinstance(zero_division, str):
```
good one, I'll add it to next version

marctorrellas commented on sklearn/metrics/classification.py:
```diff
@@ -41,6 +41,17 @@
+    if isinstance(zero_division, str):
+        if zero_division == "warn":
+            return
+    elif isinstance(zero_division, (int, float)):
```
same

marctorrellas commented on sklearn/metrics/classification.py:
```diff
@@ -1211,14 +1251,25 @@ def _prf_divide(numerator, denominator, metric, modifier, average, warn_for):
     msg = ('{0} ill-defined and being set to 0.0 {{0}} '
-           'no {1} {2}s.'.format(msg_start, modifier, axis0))
-    if len(mask) == 1:
+           'no {1} {2}s. Use ``zero_division`` parameter to control'
```
so in docstrings is `` and in strings in ` ?

marctorrellas commented on sklearn/metrics/classification.py:
```diff
@@ -1211,14 +1251,25 @@ def _prf_divide(numerator, denominator, metric, modifier, average, warn_for):
+    return result
+
+
+def _build_prf_warning_message(average, modifier, msg_start, result_size):
```
sure

marctorrellas commented on sklearn/metrics/classification.py:
```diff
@@ -1211,14 +1251,25 @@ def _prf_divide(numerator, denominator, metric, modifier, average, warn_for):
+
+
+def _build_prf_warning_message(average, modifier, msg_start, result_size):
+    axis0 = 'sample'
```
👍

marctorsoc committed:
small refactoring requested by @thomasjpfan

marctorrellas commented:
> Would `on_zero_division` be a better name?

IMHO, it's as readable as `zero_division` so I would keep the shortest one, but I have no strong opinion about this

All the rest of changes your proposed have been applied

thomasjpfan commented:
When one sets `zero_division=1` is it obvious that it means: "If the denominator is zero, the value of this metric is 1"?

The logical is "If there is zero division then do something (warn, or set 0 or 1)". My concern is how just "zero_division" does not capture the "if..." part of the statement.

Maybe `if_zero_division` is better?

jnothman commented:
I prefer on_zero_division to if_zero_division...

marctorrellas commented:
> I prefer on_zero_division to if_zero_division...

For me just `zero_division` is fine to be honest, but if I have to choose one I would go for `on_zero_division`

Here: https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_validate.html

the `error_score` has a similar behavior. Maybe we want `zero_division_score`? (IMHO I find it too long. If a user has doubts given the name she should just check the docs...)

jnothman commented:
I suppose I can reiterate that I'm fine with zero_division

marctorrellas commented:
can we have a 4th opinion on this or just take a decision on this?

thomasjpfan commented:
I am also fine with `zero_division`.

thomasjpfan commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -1508,36 +1518,92 @@ def test_precision_recall_f1_score_with_an_empty_prediction():
+    y_pred = np.zeros_like(y_true)
+
+    func = precision_recall_fscore_support
+    p, r, f, s = assert_warns(UndefinedMetricWarning, func, y_true, y_pred,
```
Can we change the assert_warns into `pytest.warns` context managers?

```py
with pytest.warns(UndefinedMetricWarning):
	p, r, f, s = precision_recall_fscore_support(y_true, y_pred, ...)
```

and the same for the other warning assertions.

marctorrellas replied:
done

marctorsoc committed:
assert_warns --> pytest.warns

thomasjpfan commented:
Thank you @marctorrellas !
